# Tarea 3: Constrained Optimization - Programación No Lineal
## Autor: Leonard David Vivas Dallos

### Instrucción: 
La actividad a realizar es la siguiente:

Elegir dos problemas aplicados relacionados con LP y QP. Realizar un documento donde se presenten estas aplicaciones y solucionarlo usando los métodos trabajados en los cursos. Se pueden usar implementaciones disponibles. Haga un análisis de los resultados obtenidos.

Se deben subir al Moodle los siguientes archivos: 

- Informe escrito 
- Códigos usados en las implementaciones

## Problema 1: Programación Lineal

### Descripción del problema

En la vida diaria, a menudo nos encontramos con problemas cotidianos que a simple vista parecen ser difíciles de resolver, pero que en realidad pueden ser resueltos mediante la formulación de un problema de programación lineal, tales como la distribución de recursos de una empresa, la minimización de gastos en un proyecto arquitectónico, entre otros. En este caso, presentaremos un problema que tiene que ver con la distribución de recursos útiles durante una tormenta de nieve en una ciudad en Canadá, problema que perfectamente puede ser aplicado para la distribución de distintos recursos en cualquier ciudad, cambiando un poco los datos o los objetivos que se manejan. 

La formulación del problema de programación lineal fue tomada de la formulación realizada por el profesor Pedro Antonio Teppa Garrán de la Universidad Metropolitana, específicamente en una presentación de varias aplicaciones de problemas y su posterior análisis y formulación del mismo como un problema de programación lineal. Así pues, el problema es el siguiente:

Una ciudad mediana en Canadá tiene dos ubicaciones en las que se mantienen reservas de sal y arena para usarlas durante las heladas y tormentas de nieve. Durante la tormenta, se distribuyen sal y arena desde estas dos ubicaciones hacia cuatro zonas diferentes de la ciudad. En ocasiones se necesita más sal y arena. Sin embargo, con frecuencia es imposible obtener provisiones adicionales durante una tormenta dado que las reservas se encuentran en una posición central distante de la ciudad. Los funcionarios de la ciudad esperan que no haya tormentas en sitios opuestos.
En una Tabla se resume el costo de abastecimiento de una tonelada de sal o arena desde cada reserva hasta cada zona de la ciudad. Además, se indican las capacidades de reserva y los niveles normales de demanda para cada zona (en toneladas)

La tabla es la siguiente:

![](Tabla.png)

El director de obras públicas se interesa en determinar el costo mínimo de distribuir las reservas de sal y arena durante una tormenta.

### Análisis y formulación del problema

Como podemos evidenciar por el enunciado, los parámetros de nuestro problema son los siguientes:
- Costo de abastecimiento de una tonelada de sal o arena desde cada reserva hasta cada zona de la ciudad.
- Capacidades de reserva.
- Niveles normales de demanda para cada zona (en toneladas).

Nuestro objetivo será entonces minimizar el costo de distribuir las reservas de sal y arena durante una tormenta. Ahora bien, lo que sigue es formular el problema como un problema de programación lineal. Para ello, definimos las siguientes variables de decisión:

- $x_{ij}$: Cantidad de toneladas de sal o arena que se distribuyen desde la reserva $i$ hacia la zona $j$.

Por las formulaciones de nuestro problema, podemos notar que son 8 variables de decisión, ya que tenemos 2 reservas y 4 zonas, es decir, serán 8 incógnitas a resolver. 
Ahora bien, como lo que queremos es minimizar costos, nuestra función objetivo será la siguiente:

$$f = 2x_{11} + 3x_{12} + 1.5x_{13} + 2.5x_{14} + 4x_{21} + 3.5x_{22} + 2.5x_{23} + 3x_{24}$$

donde f es el costo total de distribuir las reservas de sal y arena durante una tormenta y está en dólares.

Además, por las condiciones del problema nos encontramos con distintas restricciones, en términos de las capacidades máximas que pueden ofertar cada una de las reservas y las demandas de cada una de las zonas. Por lo tanto, las restricciones serán las siguientes:

**Restricciones de capacidad de reserva:**

- $x_{11} + x_{12} + x_{13} + x_{14} \leq 900$ - Corresponde a la capacidad de la reserva 1.
- $x_{21} + x_{22} + x_{23} + x_{24} \leq 750$ - Corresponde a la capacidad de la reserva 2.

**Restricciones de demanda de zona:**

- $x_{11} + x_{21} = 300$ - Corresponde a la demanda de la zona 1.
- $x_{12} + x_{22} = 450$ - Corresponde a la demanda de la zona 2.
- $x_{13} + x_{23} = 500$ - Corresponde a la demanda de la zona 3.
- $x_{14} + x_{24} = 350$ - Corresponde a la demanda de la zona 4.

**Restricciones de no negatividad:**

- $x_{ij} \geq 0$ para todo $i$ y $j$.


Así pues, finalmente nuestro problema de programación lineal será el siguiente:

$$\text{Min } f = 2x_{11} + 3x_{12} + 1.5x_{13} + 2.5x_{14} + 4x_{21} + 3.5x_{22} + 2.5x_{23} + 3x_{24}$$
$$\text{Sujeto a: }$$
$$x_{11} + x_{12} + x_{13} + x_{14} \leq 900$$
$$x_{21} + x_{22} + x_{23} + x_{24} \leq 750$$
$$x_{11} + x_{21} = 300$$
$$x_{12} + x_{22} = 450$$
$$x_{13} + x_{23} = 500$$
$$x_{14} + x_{24} = 350$$
$$x_{11}, x_{12}, x_{13}, x_{14}, x_{21}, x_{22}, x_{23}, x_{24} \geq 0$$

### Solución del problema

Una vez tenemos formulado nuestro problema, solo resta resolverlo. Para ello, usaremos dos métodos distintos: el método de la librería Scipy y el método de la librería PuLP. Se realizará un análisis de cada uno de estos y posterior a esto se compararán los resultados obtenidos.

#### Método de la librería Scipy

In [204]:
# Importamos las librerías necesarias
from scipy.optimize import linprog

In [205]:
# Primero, deberemos definir nuestra función objetivo y las restricciones
funcion_objetivo = [2, 3, 1.5, 2.5, 4, 3.5, 2.5, 3] # Cada uno de los coeficientes de la función objetivo

# Restricciones
# Restricciones de desigualdad (capacidad de reserva)
lado_izq_desigualdad = [[1, 1, 1, 1, 0, 0, 0, 0],
                        [0, 0, 0, 0, 1, 1, 1, 1]] # Coeficientes de las restricciones de desigualdad

lado_derecho_desigualdad = [900,
                           750] # Lado derecho de las restricciones de desigualdad

# Restricciones de igualdad (demanda de zona)
lado_izq_igualdad = [[1, 0, 0, 0, 1, 0, 0, 0],
                     [0, 1, 0, 0, 0, 1, 0, 0],
                     [0, 0, 1, 0, 0, 0, 1, 0],
                     [0, 0, 0, 1, 0, 0, 0, 1]] # Coeficientes de las restricciones de igualdad

lado_derecho_igualdad = [300,
                         450,
                         500,
                         350] # Lado derecho de las restricciones de igualdad

In [206]:
# Definimos los límites de las variables (Se definen por buenas prácticas, ya que por defecto las variables son no negativas)
limite_variables = [(0, float("inf")) for i in range(8)] # Las variables son no negativas

In [207]:
# Resolvemos el problema
resultado_s = linprog(c=funcion_objetivo, A_ub=lado_izq_desigualdad, b_ub=lado_derecho_desigualdad,
                      A_eq=lado_izq_igualdad, b_eq=lado_derecho_igualdad, bounds=limite_variables, 
                      method="interior-point")

C:\Users\David\AppData\Local\Temp\ipykernel_29768\3639968613.py:2: DeprecationWarning: `method='interior-point'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  resultado_s = linprog(c=funcion_objetivo, A_ub=lado_izq_desigualdad, b_ub=lado_derecho_desigualdad,


In [208]:
# Mostramos los resultados
print("Solución del problema de programación lineal usando Scipy:")
print(resultado_s)

Solución del problema de programación lineal usando Scipy:
 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: 3924.9999611675694
       x: [ 3.000e+02  1.320e+01  5.000e+02  8.680e+01  1.106e-06
            4.368e+02  3.328e-06  2.632e+02]
     nit: 6


In [209]:
# Mostrando un poco mejor los resultados tenemos
print("Solución del problema de programación lineal usando Scipy:")
print("Costo total: ", resultado_s.fun)
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 1: ", resultado_s.x[0])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 2: ", resultado_s.x[1])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 3: ", resultado_s.x[2])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 4: ", resultado_s.x[3])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 1: ", resultado_s.x[4])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 2: ", resultado_s.x[5])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 3: ", resultado_s.x[6])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 4: ", resultado_s.x[7])

print("Optimización finalizada: ", resultado_s.success)
print("Cantidad de iteraciones: ", resultado_s.nit)

Solución del problema de programación lineal usando Scipy:
Costo total:  3924.9999611675694
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 1:  299.9999956757198
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 2:  13.199370580672007
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 3:  499.9999912931002
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 4:  86.80063273557427
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 1:  1.1058467363375432e-06
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 2:  436.80062458087764
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 3:  3.3284432350422902e-06
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 4:  263.1993635059869
Optimiza

#### Método de la librería PuLP

In [210]:
# Importamos las librerías necesarias
from pulp import LpMinimize, LpVariable, LpStatus, LpProblem

In [211]:
# Creamos el problema
problema = LpProblem(name="Distribución de sal y arena", sense=LpMinimize)

In [212]:
# Definimos las variables de decisión
x11 = LpVariable(name="x11", lowBound=0)
x12 = LpVariable(name="x12", lowBound=0)
x13 = LpVariable(name="x13", lowBound=0)
x14 = LpVariable(name="x14", lowBound=0)
x21 = LpVariable(name="x21", lowBound=0)
x22 = LpVariable(name="x22", lowBound=0)
x23 = LpVariable(name="x23", lowBound=0)
x24 = LpVariable(name="x24", lowBound=0)

In [213]:
# Añadimos las restricciones al modelo
problema += (x11 + x12 + x13 + x14 <= 900, "capacidad_reserva1")
problema += (x21 + x22 + x23 + x24 <= 750, "capacidad_reserva2")
problema += (x11 + x21 == 300, "demanda_zona1")
problema += (x12 + x22 == 450, "demanda_zona2")
problema += (x13 + x23 == 500, "demanda_zona3")
problema += (x14 + x24 == 350, "demanda_zona4")

In [214]:
# Definimos la función objetivo
problema += 2*x11 + 3*x12 + 1.5*x13 + 2.5*x14 + 4*x21 + 3.5*x22 + 2.5*x23 + 3*x24

In [215]:
# Veamos el problema
print(problema)

Distribución_de_sal_y_arena:
MINIMIZE
2*x11 + 3*x12 + 1.5*x13 + 2.5*x14 + 4*x21 + 3.5*x22 + 2.5*x23 + 3*x24 + 0.0
SUBJECT TO
capacidad_reserva1: x11 + x12 + x13 + x14 <= 900

capacidad_reserva2: x21 + x22 + x23 + x24 <= 750

demanda_zona1: x11 + x21 = 300

demanda_zona2: x12 + x22 = 450

demanda_zona3: x13 + x23 = 500

demanda_zona4: x14 + x24 = 350

VARIABLES
x11 Continuous
x12 Continuous
x13 Continuous
x14 Continuous
x21 Continuous
x22 Continuous
x23 Continuous
x24 Continuous


In [216]:
# Resolvemos el problema
status = problema.solve()

In [217]:
# Mostramos los resultados
print("Solución del problema de programación lineal usando PuLP:")
print(status) # Si el status es 1, entonces la solución es óptima

Solución del problema de programación lineal usando PuLP:
1


In [218]:
# Mostramos los resultados
print("Solución del problema de programación lineal usando PuLP:")
print("Status: ", LpStatus[status])
print("Costo total: ", problema.objective.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 1: ", x11.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 2: ", x12.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 3: ", x13.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 4: ", x14.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 1: ", x21.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 2: ", x22.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 3: ", x23.value())
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 4: ", x24.value())


Solución del problema de programación lineal usando PuLP:
Status:  Optimal
Costo total:  3925.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 1:  300.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 2:  0.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 3:  500.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 4:  100.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 1:  0.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 2:  450.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 3:  0.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 4:  250.0


In [219]:
# Impresión del modelo usado para analizarlo
print(problema.solver)

#### Análisis de los resultados

Como pudimos notar, los resultados obtenidos por ambas librerías son distintos en lo que respecta a los valores de las variables de decisión, pero el costo total obtenido es aproximadamente el mismo. En una mirada somera podría parecer raro pero se deduce que esto se genera por la forma en que cada librería resuelve el problema. En el caso de Scipy, se resuelve el problema mediante el método del punto interior, mientras que en PuLP se resuelve mediante el método simplex primal o dual, por la manera en que se conecta la librería con la API. Por lo tanto, es normal que los resultados sean distintos, pero el costo total obtenido es el mismo, lo cual nos indica que ambos métodos son correctos y que la solución es óptima en ambos casos. Analicemos un poco más a detalle los resultados obtenidos por cada librería:

- **Costo total:** El costo total obtenido por ambas librerías es de aproximadamente $3925, este valor difiere un poco en las últimas cifras decimales, pero que resultan no ser significativas. Esto nos sugiere que nuestra función objetivo tiene varias soluciones óptimas, lo que sugiere que es un problema "degenerado". Esto deriva en cambios en las variables de decisión, que veremos a continuación.
- **Variables de decisión:** Las variables de decisión obtenidas por ambas librerías son distintas, a pesar de que generan el mismo costo total. Esto se debe a la forma en que cada librería resuelve el problema, como mencionamos anteriormente.

Para una mayor seguridad y comprensión de que los resultados son correctos, podríamos intentar resolver el problema usando Scipy con otro método, como el método simplex, para ver si obtenemos los mismos resultados que con PuLP, ya que apriori esta última lo resuelve mediante el método simplex. Esto se hace a pesar de que no es cuestión del curso la aplicación del método simplex, pero para efectos de comparación y análisis de los resultados, sería interesante hacerlo.

In [220]:
# Resolviendo el problema con Scipy y el método simplex
resultado_s_simplex = linprog(c=funcion_objetivo, A_ub=lado_izq_desigualdad, b_ub=lado_derecho_desigualdad,
                              A_eq=lado_izq_igualdad, b_eq=lado_derecho_igualdad, bounds=limite_variables, 
                              method="simplex")

# Mostramos los resultados
print("Solución del problema de programación lineal usando Scipy y el método simplex:")
print("Costo total: ", resultado_s_simplex.fun)
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 1: ", resultado_s_simplex.x[0])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 2: ", resultado_s_simplex.x[1])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 3: ", resultado_s_simplex.x[2])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 4: ", resultado_s_simplex.x[3])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 1: ", resultado_s_simplex.x[4])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 2: ", resultado_s_simplex.x[5])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 3: ", resultado_s_simplex.x[6])
print("Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 4: ", resultado_s_simplex.x[7])

print("Optimización finalizada: ", resultado_s_simplex.success)
print("Cantidad de iteraciones: ", resultado_s_simplex.nit)


Solución del problema de programación lineal usando Scipy y el método simplex:
Costo total:  3925.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 1:  300.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 2:  0.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 3:  500.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 1 hacia la zona 4:  100.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 1:  0.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 2:  450.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 3:  0.0
Cantidad de toneladas de sal o arena que se distribuyen desde la reserva 2 hacia la zona 4:  250.0
Optimización finalizada:  True
Cantidad de iteraciones:  10


C:\Users\David\AppData\Local\Temp\ipykernel_29768\2417028996.py:2: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  resultado_s_simplex = linprog(c=funcion_objetivo, A_ub=lado_izq_desigualdad, b_ub=lado_derecho_desigualdad,


Como podemos corroborar, la diferencia en la determinación del método a usar para resolver nuestro problema de optimización es de vital importancia para obtener una respuesta u otra. En este caso, obtenemos una minimización del costo total, pero obtenemos valores de decisión distintos para cada uno de los casos. Así pues, se puede concluir que es de vital importancia la elección del método a usar para resolver un problema de optimización, ya que esto puede llevar a obtener resultados distintos, aunque el objetivo sea el mismo.

Ahora bien, tomando en cuenta esta última solución del problema son Scipy que nos retorna la cantidad de iteraciones usadas para resolver nuestro problema, podemos deducir que para nuestro caso específico, el método de punto interior llega a la solución más rápido que el método simplex, con una diferencia de 4 iteraciones, lo cual a pesar de ser bastante poco para nuestro caso, puede llegar a ser importante en problemas a gran escala, donde la cantidad de iteraciones puede ser mucho mayor, aunque esto es un análisis que se hace con respecto a las condiciones específicas de nuestro problema, por lo que no se puede generalizar para todos los problemas de optimización.

Finalmente, en términos de significado pragmático en nuestro problema, podríamos decir que el segundo de los métodos es mucho mejor, pues nos resulta en valores de la función que son casi enteros, sin muchos decimales. Esto puede llegar a ser útil en el uso de los resultados, ya que se pueden interpretar de una manera más sencilla y rápida, sin necesidad de hacer redondeos o ajustes a los resultados obtenidos.

## Problema 2: Programación Cuadrática

### Descripción del problema

Uno de los problemas más interesantes que a mi parecer, pueden ser aplicados a la programación cuadrática, se basa en una situación cotidiana a la que se ven enfrentados miles de economistas y empresarios como tal, cuando están en la tarea de analizar y encontrar un portafolio de inversión que maximice los beneficios, pero que a su vez minimice los riesgos.

En este caso, nos hemos basado en el documento "Numerical solution of a general interval quadratic programming model for portfolio selection" escrito por Wang J, He F, Shi X, y citado de la siguiente manera: Wang J, He F, Shi X (2019) Numerical solution of a general interval quadratic programming model for portfolio selection. PLoS ONE 14(3): e0212913. https://doi.org/10.1371/journal.pone.0212913

En este documento, se realiza el análisis de las soluciones numéricas basadas en un modelo de programación cuadrática para la selección de portafolios de inversión, para este fin analiza el modelo desarrollado por Markowitz, con sus propias consideraciones para obtener un modelo general a optimizar, y más aún, se analizan posteriores ejemplos prácticos que serán usados para el fin de estudio en este Notebook.

El modelo general para la selección de portafolios se describe como sigue:

Supongamos que hay $n$ tipos de activos para que los inversores selecccionen. Basados en el modelo de media-varianza, los inversores pretenden minimizar el riesgo de la cartera $f(x) = x^T \hat{Q} x = \sum_{i=1}^{n} \sum_{j=1}^{n} q_{ij} x_i x_j$ bajo condiciones de rendimientos fijos $\sum_{i=1}^{n} \hat{R}_i x_i - \sum_{i=1}^{n} c_i x_i \geq \hat{R}_0$ y tasa de rotación $\sum_{i=1}^{n} \hat{l}_i x_i \geq \hat{l}_0$. Establecemos entonces un modelo de programación cuadrática general para la selección de portafolios de inversión:

$$\text{Min } f(x) = x^T \hat{Q} x = \sum_{i=1}^{n} \sum_{j=1}^{n} \hat{q}_{ij} x_i x_j$$
$$\text{Sujeto a: }$$
$$\sum_{i=1}^{n} \hat{R}_i x_i - \sum_{i=1}^{n} c_i x_i \geq \hat{R}_0$$
$$\sum_{i=1}^{n} \hat{l}_i x_i \geq \hat{l}_0$$
$$\sum_{i=1}^{n} x_i = 1$$
$$x_i \geq 0, i = 1, 2, ..., n$$

donde $c_i$ es el costo de transacción de activo $i$, $x_i$ es la fracción de inversión en activo $i$, $\hat{r}_i$ es el retorno del activo $i$. $\hat{R}_i$ y $\hat{l}_i$ son el retorno esperado y las tasas de rotación de activo $i$, respectivamente. $\hat{Q} = (q_{ij})$ es la matriz de covarianza de los rendimientos de los activos, donde se asume que $\hat{Q}$ es semidefinida positiva. Como $\hat{r}_i$ y $\hat{R}_i$ son inciertos, se consideran intervalos para ellos, es decir, $\hat{r}_i = [\underline{r}_i, \overline{r}_i]$ y $\hat{R}_i = E\hat{r}_i = [\underline{R}_i, \overline{R}_i]$, $\hat{q}_{ij} = [\underline{q}_{ij}, \overline{q}_{ij}]$, y $\hat{l}_i = [\underline{l}_i, \overline{l}_i]$. Así, solucionando $x = (x_1, x_2, ..., x_n)^T$, en el modelo, se obtiene un portafoilio de inversión.

Para resolver este problema, debemos obtener un rango para la función minimizada dada la incertidumbre de los rendimientos, manejando un caso pesimista y un caso optimista. Ahora bien, en el documento que se menciona anteriormente se hace un análisis de un método primal-dual del Lagrangiano para resolver dichas cotas que estamos buscando y que de hecho, obtiene unos resultados bastante buenos y que mejoran en cierta medida otros previamente analizados. Sin embargo, como este no fue un método en el que se ahondó mucho en clase, se hará uso del otro método descrito en el documento, un método convencional que pasa el modelo anterior a un modelo determinista, simplemente dividiendo entre las dos cotas de los intervalos que se brindan y así se obtiene la solución. Así pues, los dos modelos deterministas que se obtienen son los siguientes:

**Modelo optimista:**

$$\text{Min } f^{U}(x) = x^T Q x = \sum_{i=1}^{n} \sum_{j=1}^{n} \overline{q}_{ij} x_i x_j$$
$$\text{Sujeto a: }$$
$$\sum_{i=1}^{n} \underline{R}_i x_i - \sum_{i=1}^{n} c_i x_i \geq \overline{R}_0$$
$$\sum_{i=1}^{n} \underline{l}_i x_i \geq \overline{l}_0$$
$$\sum_{i=1}^{n} x_i = 1$$
$$x_i \geq 0, i = 1, 2, ..., n$$

**Modelo pesimista:**

$$\text{Min } f^{L}(x) = x^T Q x = \sum_{i=1}^{n} \sum_{j=1}^{n} \underline{q}_{ij} x_i x_j$$
$$\text{Sujeto a: }$$
$$\sum_{i=1}^{n} \overline{R}_i x_i - \sum_{i=1}^{n} c_i x_i \geq \underline{R}_0$$
$$\sum_{i=1}^{n} \overline{l}_i x_i \geq \underline{l}_0$$
$$\sum_{i=1}^{n} x_i = 1$$
$$x_i \geq 0, i = 1, 2, ..., n$$

Así pues, resolveremos ambos modelos para encontrar un rango de decisión de nuestra función objetivo general.

### Caso de estudio

En el documento mencionado, se describe el siguiente caso de estudio como ejemplo 1:

De acuerdo con "Deng X, Zhao J. F. & Li R. J. An investment portfolio selection model based on the satisfaction index of interval inequality. Statistics and Decision, 2010 (22): 145–147.", seleccionamos 3 tipos de valores desde Abril 2005 hasta Marzo de 2009 en Guangzhou Holdings, Shanghai Airport, Minmetals Development. Considerando el cierre de precio mensual y la tasa de rotación, calculamos los intervalos de tasa de retorno esperada, intervalos de riesgo de varianza y covarianza e intervalos de tasa de rotación de los tres activos.

Los intervalos de tasa de retorno esperada son los siguientes:

- $\hat{R}_1 = [-0.02972, 0.02196]$
- $\hat{R}_2 = [-0.02259, 0.01803]$
- $\hat{R}_3 = [0.00282, 0.06566]$

Los intervalos de riesgo de varianza y covarianza son los siguientes:

- $\hat{q}_{11} = [0.0204, 0.0289]$
- $\hat{q}_{12} = \hat{q}_{21} = [0.0174, 0.0212]$
- $\hat{q}_{13} = \hat{q}_{31} = [0.0213, 0.025]$
- $\hat{q}_{22} = [0.0179, 0.0269]$
- $\hat{q}_{23} = \hat{q}_{32} = [0.0164, 0.0320]$
- $\hat{q}_{33} = [0.0417, 0.0590]$

Los intervalos de tasa de rotación son los siguientes:

- $\hat{l}_1 = [0.2724, 0.4067]$
- $\hat{l}_2 = [0.2211, 0.2569]$
- $\hat{l}_3 = [0.7688, 1.2066]$

Suponga que las tasas de costos de transacción de los tres activos son $c_1 = 0.00015$, $c_2 = 0.00025$, $c_3 = 0.0002$. El intervalo de retorno mínimo esperado de los tres activos fue establecida como $\hat{R}_0 = [0.001, 0.0025]$, es decir, la tasa de rendimiento esperado por los inversores es $0.001$ en el caso pesimista y $0.0025$ en el caso optimista. El intervalo de tasa de rotación esperada de los tres activos fue establecido como $\hat{l}_0 = [0.40, 0.60]$, es decir $0.40$ en el caso pesimista, y $0.60$ en el caso optimista.

#### Solución del problema para el caso pesimista

Nuestro modelo a resolver para el caso pesimista es el siguiente:

$$\text{Min } f^{L}(x) = x^T Q x = \sum_{i=1}^{3} \sum_{j=1}^{3} \underline{q}_{ij} x_i x_j$$
$$\text{Sujeto a: }$$
$$\sum_{i=1}^{3} \overline{R}_i x_i - \sum_{i=1}^{3} c_i x_i \geq \underline{R}_0$$
$$\sum_{i=1}^{3} \overline{l}_i x_i \geq \underline{l}_0$$
$$\sum_{i=1}^{3} x_i = 1$$
$$x_i \geq 0, i = 1, 2, 3$$

Donde los valores dados son los siguientes:

- $\underline{q}_{11} = 0.0204$
- $\underline{q}_{12} = \underline{q}_{21} = 0.0174$
- $\underline{q}_{13} = \underline{q}_{31} = 0.0213$
- $\underline{q}_{22} = 0.0179$
- $\underline{q}_{23} = \underline{q}_{32} = 0.0164$
- $\underline{q}_{33} = 0.0417$
- $\overline{R}_1 = 0.02196$
- $\overline{R}_2 = 0.01803$
- $\overline{R}_3 = 0.06566$
- $c_1 = 0.00015$
- $c_2 = 0.00025$
- $c_3 = 0.0002$
- $\underline{R}_0 = 0.001$
- $\overline{l}_1 = 0.4067$
- $\overline{l}_2 = 0.2569$
- $\overline{l}_3 = 1.2066$
- $\underline{l}_0 = 0.40$
- $n = 3$


In [221]:
# Importamos las librerías necesarias
import numpy as np
from qpsolvers import solve_qp

# Definimos los valores de los parámetros
Q_p = np.array([[0.0204, 0.0174, 0.0213],
              [0.0174, 0.0179, 0.0164],
              [0.0213, 0.0164, 0.0417]])

# Definimos la matriz de las restricciones de desigualdad
# Las primeras serán entonces (R_i - c_i)x_i >= R_0, las filas siguientes serán las restricciones de rotación de activos (l_i)x_i => l_0
G_p = np.array([[-(0.02196 - 0.00015), -(0.01803 - 0.00025), -(0.06566 - 0.0002)],
              [-0.4067, -0.2569, -1.2066]])

# Definimos el vector solución de las restricciones de desigualdad
h_p = np.array([-0.001, -0.40])

# Definimos la matriz de las restricciones de igualdad
# La restricción de igualdad será simplemente la suma de las fracciones de inversión en los activos
A_p = np.array([[1.0, 1.0, 1.0]])

# Definimos el vector solución de las restricciones de igualdad
b_p = np.array([1.0])

# Definimos el vector q que es la parte lineal (en nuestro caso no hay)
q_p = np.zeros(3)

# Definimos el límite inferior de las variables
lb_p = np.zeros(3)

In [222]:
# Verificación de la matriz Q
eigenvalues_p = np.linalg.eigvals(Q_p)
print("Autovalores de Q:", eigenvalues_p)

Autovalores de Q: [0.06608735 0.00139032 0.01252233]


La última verificación se hizo para asegurarnos de que la matriz Q sea semidefinida positiva, ya que es un requisito para que el problema de programación cuadrática sea resoluble. En este caso, los autovalores son todos positivos, por lo que podemos proceder a resolver el problema.

##### Solución del problema por el método Interior Point de la librería qpsolvers, con el solver cvxopt

In [223]:
# Resolvemos el problema
x_pi = solve_qp(Q_p, q_p, G_p, h_p, A_p, b_p, lb_p, solver="cvxopt", verbose=True)

     pcost       dcost       gap    pres   dres
 0:  9.9187e-03 -1.2542e+00  1e+00  6e-17  4e+00
 1:  9.9122e-03 -4.9916e-03  1e-02  1e-16  5e-02
 2:  9.5824e-03  8.0552e-03  2e-03  7e-17  4e-03
 3:  9.0979e-03  8.8846e-03  2e-04  6e-17  2e-18
 4:  9.0313e-03  9.0180e-03  1e-05  2e-16  7e-18
 5:  9.0269e-03  9.0255e-03  1e-06  1e-16  1e-17
 6:  9.0267e-03  9.0266e-03  7e-08  1e-16  1e-16
Optimal solution found.


In [224]:
# Mostramos los resultados
print("Solución del problema de programación cuadrática para el caso pesimista:")
print("Fracción de inversión en el activo 1: ", x_pi[0])
print("Fracción de inversión en el activo 2: ", x_pi[1])
print("Fracción de inversión en el activo 3: ", x_pi[2])

Solución del problema de programación cuadrática para el caso pesimista:
Fracción de inversión en el activo 1:  0.03593381634312439
Fracción de inversión en el activo 2:  0.8190546406401066
Fracción de inversión en el activo 3:  0.14501154301676908


In [225]:
# Verificación de las restricciones de desigualdad
print("Restricciones de desigualdad:")
print("Ri*xi - ci*xi >= R0: ", (0.02196 - 0.00015)*x_pi[0] + (0.01803 - 0.00025)*x_pi[1] + (0.06566 - 0.0002)*x_pi[2] >= 0.001)

Restricciones de desigualdad:
Ri*xi - ci*xi >= R0:  True


In [226]:
# Verificación de las restricciones de igualdad
print("Restricciones de igualdad:")
print("Suma de las fracciones de inversión: ", np.sum(x_pi))


Restricciones de igualdad:
Suma de las fracciones de inversión:  1.0


In [253]:
# Verificación de las restricciones de rotación
print("Restricciones de rotación:")
print("li*xi >= l0: ", 0.4067*x_pi[0] + 0.2569*x_pi[1] + 1.2066*x_pi[2] >= 0.40)

Restricciones de rotación:
li*xi >= l0:  True


In [228]:
# Valor de la función objetivo
f_pi = x_pi.T @ Q_p @ x_pi
print("Valor de la función objetivo: ", f_pi)


Valor de la función objetivo:  0.018053387443620693


##### Solución del problema por el método Active Set de la librería qpsolvers con el solver daqp

In [229]:
# Resolvemos el problema
x_pa = solve_qp(Q_p, q_p, G_p, h_p, A_p, b_p, lb_p, solver="daqp", verbose=True)

In [230]:
# Mostramos los resultados
print("Solución del problema de programación cuadrática para el caso pesimista:")
print("Fracción de inversión en el activo 1: ", x_pa[0])
print("Fracción de inversión en el activo 2: ", x_pa[1])
print("Fracción de inversión en el activo 3: ", x_pa[2])


Solución del problema de programación cuadrática para el caso pesimista:
Fracción de inversión en el activo 1:  0.035194564031902624
Fracción de inversión en el activo 2:  0.8196776542391081
Fracción de inversión en el activo 3:  0.1451277817289892


In [231]:
# Verificación de las restricciones de desigualdad
print("Restricciones de desigualdad:")
print("Ri*xi - ci*xi >= R0: ", (0.02196 - 0.00015)*x_pa[0] + (0.01803 - 0.00025)*x_pa[1] + (0.06566 - 0.0002)*x_pa[2] >= 0.001)

Restricciones de desigualdad:
Ri*xi - ci*xi >= R0:  True


In [255]:
# Verificación de las restricciones de igualdad
print("Restricciones de igualdad:")
print("Suma de las fracciones de inversión: ", np.sum(x_pa))

Restricciones de igualdad:
Suma de las fracciones de inversión:  0.9999999999999999


In [254]:
# Verificación de las restricciones de rotación
print("Restricciones de rotación:")
print("li*xi >= l0: ", 0.4067*x_pa[0] + 0.2569*x_pa[1] + 1.2066*x_pa[2] >= 0.40)

Restricciones de rotación:
li*xi >= l0:  True


In [234]:
# Valor de la función objetivo
f_pa = x_pa.T @ Q_p @ x_pa
print("Valor de la función objetivo: ", f_pa)

Valor de la función objetivo:  0.018053384205928863


#### Solución del problema para el caso optimista

Nuestro modelo a resolver para el caso optimista es el siguiente:

$$\text{Min } f^{U}(x) = x^T Q x = \sum_{i=1}^{3} \sum_{j=1}^{3} \overline{q}_{ij} x_i x_j$$
$$\text{Sujeto a: }$$
$$\sum_{i=1}^{3} \underline{R}_i x_i - \sum_{i=1}^{3} c_i x_i \geq \overline{R}_0$$
$$\sum_{i=1}^{3} \underline{l}_i x_i \geq \overline{l}_0$$
$$\sum_{i=1}^{3} x_i = 1$$
$$x_i \geq 0, i = 1, 2, 3$$

Donde los valores dados son los siguientes:

- $\overline{q}_{11} = 0.0289$
- $\overline{q}_{12} = \overline{q}_{21} = 0.0212$
- $\overline{q}_{13} = \overline{q}_{31} = 0.025$
- $\overline{q}_{22} = 0.0269$
- $\overline{q}_{23} = \overline{q}_{32} = 0.0320$
- $\overline{q}_{33} = 0.0590$
- $\underline{R}_1 = -0.02972$
- $\underline{R}_2 = -0.02259$
- $\underline{R}_3 = 0.00282$
- $c_1 = 0.00015$
- $c_2 = 0.00025$
- $c_3 = 0.0002$
- $\overline{R}_0 = 0.0025$
- $\underline{l}_1 = 0.2724$
- $\underline{l}_2 = 0.2211$
- $\underline{l}_3 = 0.7688$
- $\overline{l}_0 = 0.60$
- $n = 3$

In [235]:
# Importamos las librerías necesarias
import numpy as np
from qpsolvers import solve_qp

# Definimos los valores de los parámetros
Q_o = np.array([[0.0289, 0.0212, 0.025],
              [0.0212, 0.0269, 0.0320],
              [0.025, 0.0320, 0.0590]])

# Definimos la matriz de las restricciones de desigualdad
# Las primeras serán entonces (R_i - c_i)x_i >= R_0, las filas siguientes serán las restricciones de rotación de activos (l_i)x_i >= l_0
G_o = np.array([[-(-0.02972 - 0.00015), -(-0.02259 - 0.00025), -(0.00282 - 0.0002)],
                [-0.2724, -0.2211, -0.7688]])

# Definimos el vector solución de las restricciones de desigualdad
h_o = np.array([-0.0025, -0.60])

# Definimos la matriz de las restricciones de igualdad
# La restricción de igualdad será simplemente la suma de las fracciones de inversión en los activos
A_o = np.array([[1.0, 1.0, 1.0]])

# Definimos el vector solución de las restricciones de igualdad
b_o = np.array([1.0])

# Definimos el vector q que es la parte lineal (en nuestro caso no hay)
q_o = np.zeros(3)

# Definimos el límite inferior de las variables
lb_o = np.zeros(3)

In [236]:
# Verificación de la matriz Q
eigenvalues_o = np.linalg.eigvals(Q_o)
print("Autovalores de Q:", eigenvalues_o)

Autovalores de Q: [0.09476341 0.01530335 0.00473324]


Nuevamente, la última verificación se hizo para asegurarnos de que la matriz Q sea semidefinida positiva, ya que es un requisito para que el problema de programación cuadrática sea resoluble. En este caso, los autovalores son todos positivos, por lo que podemos proceder a resolver el problema.

##### Solución del problema por el método Interior Point de la librería qpsolvers, con el solver cvxopt

In [237]:
# Resolvemos el problema
x_oi = solve_qp(Q_o, q_o, G_o, h_o, A_o, b_o, lb_o, solver="cvxopt", verbose=True)

     pcost       dcost       gap    pres   dres
 0:  1.5662e-02 -7.6294e-01  8e+00  3e+00  3e+00
 1:  2.2480e-02 -4.7785e-01  6e-01  4e-02  6e-02
 2:  2.2923e-02 -1.6725e-01  2e-01  2e-02  2e-02
 3:  2.8876e-02 -6.7229e-02  2e-01  1e-02  1e-02
 4:  2.9436e-02  2.5089e-02  7e-03  3e-04  4e-04
 5:  2.9464e-02  2.8646e-02  8e-04  9e-07  1e-06
 6:  2.9456e-02  2.9363e-02  9e-05  1e-07  1e-07
 7:  2.9374e-02  2.9335e-02  4e-05  2e-08  3e-08
 8:  2.9374e-02  2.9372e-02  1e-06  7e-10  9e-10
 9:  2.9373e-02  2.9373e-02  6e-08  8e-12  1e-11
Optimal solution found.


In [238]:
# Mostramos los resultados
print("Solución del problema de programación cuadrática para el caso optimista:")
print("Fracción de inversión en el activo 1: ", x_oi[0])
print("Fracción de inversión en el activo 2: ", x_oi[1])
print("Fracción de inversión en el activo 3: ", x_oi[2])

Solución del problema de programación cuadrática para el caso optimista:
Fracción de inversión en el activo 1:  1.371859633388854e-05
Fracción de inversión en el activo 2:  0.00469547229183114
Fracción de inversión en el activo 3:  0.995290809111835


In [239]:
# Verificación de las restricciones de desigualdad
print("Restricciones de desigualdad:")
print("Ri*xi - ci*xi >= R0: ", (-0.02972 - 0.00015)*x_oi[0] + (-0.02259 - 0.00025)*x_oi[1] + (0.00282 - 0.0002)*x_oi[2] >= 0.0025)

Restricciones de desigualdad:
Ri*xi - ci*xi >= R0:  True


In [240]:
# Verificación de las restricciones de igualdad
print("Restricciones de igualdad:")
print("Suma de las fracciones de inversión: ", np.sum(x_oi))

Restricciones de igualdad:
Suma de las fracciones de inversión:  1.0


In [241]:
# Verificación de las restricciones de rotación
print("Restricciones de rotación:")
print("li*xi >= l0: ", 0.2724*x_oi[0] + 0.2211*x_oi[1] + 0.7688*x_oi[2] >= 0.60)

Restricciones de rotación:
li*xi >= l0:  True


In [242]:
# Valor de la función objetivo
f_oi = x_oi.T @ Q_o @ x_oi
print("Valor de la función objetivo: ", f_oi)

Valor de la función objetivo:  0.05874599746706755


##### Solución del problema por el método Active Set de la librería qpsolvers con el solver daqp

In [243]:
# Resolvemos el problema
x_oa = solve_qp(Q_o, q_o, G_o, h_o, A_o, b_o, lb_o, solver="daqp", verbose=True)

In [244]:
# Mostramos los resultados
print("Solución del problema de programación cuadrática para el caso optimista:")
print("Fracción de inversión en el activo 1: ", x_oa[0])
print("Fracción de inversión en el activo 2: ", x_oa[1])
print("Fracción de inversión en el activo 3: ", x_oa[2])

Solución del problema de programación cuadrática para el caso optimista:
Fracción de inversión en el activo 1:  3.535190224033393e-17
Fracción de inversión en el activo 2:  0.004713275726629762
Fracción de inversión en el activo 3:  0.9952867242733703


In [245]:
# Verificación de las restricciones de desigualdad
print("Restricciones de desigualdad:")
print("Ri*xi - ci*xi >= R0: ", (-0.02972 - 0.00015)*x_oa[0] + (-0.02259 - 0.00025)*x_oa[1] + (0.00282 - 0.0002)*x_oa[2] >= 0.0025)

Restricciones de desigualdad:
Ri*xi - ci*xi >= R0:  True


In [246]:
# Verificación de las restricciones de igualdad
print("Restricciones de igualdad:")
print("Suma de las fracciones de inversión: ", np.sum(x_oa))

Restricciones de igualdad:
Suma de las fracciones de inversión:  1.0


In [247]:
# Verificación de las restricciones de rotación
print("Restricciones de rotación:")
print("li*xi >= l0: ", 0.2724*x_oa[0] + 0.2211*x_oa[1] + 0.7688*x_oa[2] >= 0.60)

Restricciones de rotación:
li*xi >= l0:  True


In [248]:
# Valor de la función objetivo
f_oa = x_oa.T @ Q_o @ x_oa
print("Valor de la función objetivo: ", f_oa)

Valor de la función objetivo:  0.05874596961856284



#### Análisis de los resultados

Lo primero que haremos es condensar la solución obtenida. La construcción de la solución está basada en juntar los dos intervalos solución obtenidos para cada uno de los modelos, de esta manera se obtiene un intervalo solución para la función objetivo, en nuestro caso, obtenemos el intervalo para nuestro portafolio de inversión.

Según nuestras soluciones, tenemos que el intervalo solución obtenido por cada uno de los métodos para el portafolio de inversión con costes de transacción y tasas de rotación, es el siguiente:

##### Intervalo obtenido por el método Interior Point (cvxopt)

In [249]:
print("Intervalo obtenido por el método Interior Point (cvxopt):")
print(f"[ {f_pi}, {f_oi} ]")

Intervalo obtenido por el método Interior Point (cvxopt):
[ 0.018053387443620693, 0.05874599746706755 ]


##### Intervalo obtenido por el método Active Set (daqp)

In [250]:
print("Intervalo obtenido por el método Active Set (daqp):")
print(f"[ {f_pa}, {f_oa} ]")

Intervalo obtenido por el método Active Set (daqp):
[ 0.018053384205928863, 0.05874596961856284 ]


Nuestro primer análisis está basado en la correctitud de las respuestas obtenidas, como podemos ver, cada uno de los métodos nos da resultados, que al compararlos directamente con los obtenidos en nuestro documento guía, resultan ser muy cercano y variantes respecto a pequeños decimales de aproximación. Así, en primer medida sabemos que los resultados obtenidos son correctos y que el método de resolución de problemas de programación cuadrática es adecuado para nuestro caso. Ahora bien, podemos ver que los intervalos obtenidos por los dos métodos implementados son muy buenos, pues son iguales si los aproximamos a 4 decimales, lo cual nos indica que ambos métodos son correctos y que la solución es óptima en ambos casos.

Ahora bien, realicemos una comparación en términos de los resultados obtenidos por los métodos para cada uno de los modelos (optimista y pesimista).

##### Caso pesimista

In [256]:
# Resultados obtenidos por el método Interior Point
print("Resultados obtenidos por el método Interior Point:")
print("Fracción de inversión en el activo 1: ", x_pi[0])
print("Fracción de inversión en el activo 2: ", x_pi[1])
print("Fracción de inversión en el activo 3: ", x_pi[2])
print("Valor de la función objetivo: ", f_pi)
print()

# Resultados obtenidos por el método Active Set
print("Resultados obtenidos por el método Active Set:")
print("Fracción de inversión en el activo 1: ", x_pa[0])
print("Fracción de inversión en el activo 2: ", x_pa[1])
print("Fracción de inversión en el activo 3: ", x_pa[2])
print("Valor de la función objetivo: ", f_pa)

Resultados obtenidos por el método Interior Point:
Fracción de inversión en el activo 1:  0.03593381634312439
Fracción de inversión en el activo 2:  0.8190546406401066
Fracción de inversión en el activo 3:  0.14501154301676908
Valor de la función objetivo:  0.018053387443620693

Resultados obtenidos por el método Active Set:
Fracción de inversión en el activo 1:  0.035194564031902624
Fracción de inversión en el activo 2:  0.8196776542391081
Fracción de inversión en el activo 3:  0.1451277817289892
Valor de la función objetivo:  0.018053384205928863


Como podemos evidenciar, las fracciones que obtienen cada uno de los dos métodos aplicados son bastante parecidas, en especial en cuanto a la fracción del activo 3, y podemos notar que los activos 1 y 2 al parecer ser reparten los decimales "restantes". Además, podemos deducir que el valor de la función objetivo obtenida por cada uno de estos métodos es muy parecido, lo que nos indica que debe ser la aproximación la que nos hace variar un poco los resultados, pero en términos de tolerancia de los métodos (por ejemplo al hacer la suma de las fracciones de los activos, computacionalmente se tomó 0.9999999999999999 como 1, lo cual a lo mejor llegó a ser la causa de esta mínima diferencia).

##### Caso optimista

In [257]:
# Resultados obtenidos por el método Interior Point
print("Resultados obtenidos por el método Interior Point:")
print("Fracción de inversión en el activo 1: ", x_oi[0])
print("Fracción de inversión en el activo 2: ", x_oi[1])
print("Fracción de inversión en el activo 3: ", x_oi[2])
print("Valor de la función objetivo: ", f_oi)
print()

# Resultados obtenidos por el método Active Set
print("Resultados obtenidos por el método Active Set:")
print("Fracción de inversión en el activo 1: ", x_oa[0])
print("Fracción de inversión en el activo 2: ", x_oa[1])
print("Fracción de inversión en el activo 3: ", x_oa[2])
print("Valor de la función objetivo: ", f_oa)

Resultados obtenidos por el método Interior Point:
Fracción de inversión en el activo 1:  1.371859633388854e-05
Fracción de inversión en el activo 2:  0.00469547229183114
Fracción de inversión en el activo 3:  0.995290809111835
Valor de la función objetivo:  0.05874599746706755

Resultados obtenidos por el método Active Set:
Fracción de inversión en el activo 1:  3.535190224033393e-17
Fracción de inversión en el activo 2:  0.004713275726629762
Fracción de inversión en el activo 3:  0.9952867242733703
Valor de la función objetivo:  0.05874596961856284


Para este caso, podemos ver mucho más claro como los resultados pueden llegar a ser mucho más parecidos, en especial en cuanto a los activos 2 y 3, los mini decimales en los que varían son compensados con el activo 1 (en este caso, a diferencia del anterior, la suma de las fracciones de los activos suma 1 en ambos casos). Ahora bien, este resultado se repite para nuestra función objetivo, ambos valores son casi iguales, difiriendo en muy pocos decimales, lo cual nos indica que ambos métodos son correctos y que la solución es óptima en ambos casos.

Finalmente, en un análisis en especial del método Interior Point, dado que el solver cvxopt nos brinda la posibilidad de ver las iteraciones, podemos ver como en ambos casos el tiempo que se demora el método en resolver el problema es bastante poco, siendo para el primer caso 6 iteraciones y para el segundo caso 9 iteraciones, esto puede sugerir que es bastante rápido el método para solucionar el modelo, sin importar el punto de origen del mismo. Este análisis no se puede realizar en específico para el método de conjunto activo, pues el solver no nos brinda estas capacidades, sin embargo, el tiempo computacional de ejecución es muy poco y bastante parecido, en ocasiones hasta más rápido que el otro método. Sin embargo, este es un análisis para nuestro caso y dados nuestros valores, no se puede generalizar para todos los problemas de programación cuadrática, pues esto es sensitivo a las matrices brindadas y a la cantidad de datos y variables obtenidas en el problema.